### PR TITLE
Filter BLACKLIST_EXTS in PluginLoader

### LINKED ISSUE #22268: Wrong file used as module
##### ISSUE TYPE

 - Feature Idea

##### ANSIBLE VERSION

```
ansible 2.3.0 (devel 7018aa3316) last updated 2017/03/03 14:56:50 (GMT -400)
```

##### OS / ENVIRONMENT

linux

##### SUMMARY

An emacs backup file `aws.py~` in the module directory is used in preference to a sibling file called `aws`.

##### STEPS TO REPRODUCE

```
mkdir ansible-bug
cd ansible-bug
cat > playbook.yaml <<EOF
---
- hosts: localhost
  connection: local

  tasks:
    - aws: {}
EOF
cat > aws <<EOF
#!/usr/bin/python
from ansible.module_utils.basic import AnsibleModule
module = AnsibleModule({})
module.exit_json(changed=False)
EOF
cat > aws.py~ <<EOF
#!/usr/bin/python
from ansible.module_utils.basic import AnsibleModule
module = AnsibleModule({})
module.exit_json(changed=True)
EOF
ansible-playbook -vvvv -i localhost, -M . playbook.yaml
```

##### EXPECTED RESULTS

`changed=0`

##### ACTUAL RESULTS

`changed=1`

```
(conf-mgmt) aaron:ansible-bug> ansible-playbook -vvvv -i localhost, -M . playbook.yaml
No config file found; using defaults
Set default localhost to localhost
Loading callback plugin default of type stdout, v2.0 from /home/aaron/Dropbox/newscred-src/ansible/lib/ansible/plugins/callback/__init__.pyc

PLAYBOOK: playbook.yaml *************************************************************************************************************
1 plays in playbook.yaml

PLAY [localhost] ********************************************************************************************************************

TASK [Gathering Facts] **************************************************************************************************************
Using module file /home/aaron/Dropbox/newscred-src/ansible/lib/ansible/modules/system/setup.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: aaron
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/aaron/.ansible/tmp/ansible-tmp-1488571336.12-276444930035913 `" && echo ansible-tmp-1488571336.12-276444930035913="` echo /home/aaron/.ansible/tmp/ansible-tmp-1488571336.12-276444930035913 `" ) && sleep 0'
<localhost> PUT /tmp/tmpb2iKjX TO /home/aaron/.ansible/tmp/ansible-tmp-1488571336.12-276444930035913/setup.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/aaron/.ansible/tmp/ansible-tmp-1488571336.12-276444930035913/ /home/aaron/.ansible/tmp/ansible-tmp-1488571336.12-276444930035913/setup.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/aaron/.ansible/tmp/ansible-tmp-1488571336.12-276444930035913/setup.py; rm -rf "/home/aaron/.ansible/tmp/ansible-tmp-1488571336.12-276444930035913/" > /dev/null 2>&1 && sleep 0'
ok: [localhost]
META: ran handlers

TASK [aws] **************************************************************************************************************************
task path: /home/aaron/scratch/ansible-bug/ansible-bug/playbook.yaml:6
Using module file /home/aaron/scratch/ansible-bug/ansible-bug/aws.py~
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: aaron
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/aaron/.ansible/tmp/ansible-tmp-1488571336.5-232883384061572 `" && echo ansible-tmp-1488571336.5-232883384061572="` echo /home/aaron/.ansible/tmp/ansible-tmp-1488571336.5-232883384061572 `" ) && sleep 0'
<localhost> PUT /tmp/tmp5HzZiD TO /home/aaron/.ansible/tmp/ansible-tmp-1488571336.5-232883384061572/aws.py~
<localhost> EXEC /bin/sh -c 'chmod u+x /home/aaron/.ansible/tmp/ansible-tmp-1488571336.5-232883384061572/ '"'"'/home/aaron/.ansible/tmp/ansible-tmp-1488571336.5-232883384061572/aws.py~'"'"' && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python '"'"'/home/aaron/.ansible/tmp/ansible-tmp-1488571336.5-232883384061572/aws.py~'"'"'; rm -rf "/home/aaron/.ansible/tmp/ansible-tmp-1488571336.5-232883384061572/" > /dev/null 2>&1 && sleep 0'
changed: [localhost] => {
    "changed": true, 
    "invocation": {
        "module_args": {}
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP **************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0   
```

Note that when I originally tried to build this test case, I called the module files `my_module` and `my_module.py~` instead of `aws` and `aws.py~`, and it failed to demonstrate the bug. My guess is that it has something to do with the way filenames are hashed by the filesystem -- ansible-playbook uses whichever one comes first in the directory listing. If that's the case then the test might work differently for someone using a different filesystem. As a variant that shouldn't depend on filesystem implementation, try deleting the file `aws`. Then ansible-playbook unambiguously uses `aws.py~` as a module, which I would argue is incorrect behavior.

### PR DESCRIPTION
**Pull request by relrod**
##### SUMMARY

(Note this revives https://github.com/ansible/ansible/pull/27235)

Change:
Rather than hardcoding .pyo and .pyc, filter on all BLACKLIST_EXTS in
the non-legacy logic of PluginLoader (_find_fq_plugin). The two harcoded
extensions are part of BLACKLIST_EXTS already and this simply adds the
rest of the blacklisted extensions to the check.

In addition, check .endswith() instead of an exact match of the suffix,
like everywhere else that uses BLACKLIST_EXTS. This allows for
blacklisting, for example, emacs's backup files which can appear after
any extension, leading to things like `foo.py~`.

Test Plan:
Ran `ansible-playbook` against a collection where a `foo.py~` module was
getting executed instead of `foo.py` which also appeared in the same
directory. `foo.py~` is no longer executed.

Tickets:
Fixes #22268
Refs #27235

Signed-off-by: Rick Elrod <rick@elrod.me>

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

plugin loader

### CODE REVIEW

**Review by nitzmahone — COMMENTED** [review]
Ugh, this extension fallback stuff is such a PITA- it really only exists to support arbitrary script/binary modules, but has "leaked" into lots of other areas where it causes problems. The ultimate solution is probably to change the base PluginLoader to default to only `*.py`, with an optional fallback blacklist for loaders that allow extension fudging (eg the module loader).

Controller plugin loaders should only allow *.py and not fall back to anything else. The discussion/comments about .pyc and .pyo being OK in certain contexts is generally wrong - the way the PluginLoader works today, we can basically never trust a binary Python module, because if it's bogus (eg wrong optimization level, wrong Python version), we're not going to go back looking for the source version, so things are just busted. Better to pretend they don't exist and rely on the PEP302 loader to implement those optimizations controller side when the module is actually loaded (which ours currently does not).

We might be able to fix the module case by duplicating the search separately at the topmost level for arbitrarily-extensioned modules *after* we've already exhausted the search for the specific extension reqeusted, which would be the only thing that should consult the extension blacklist.

I think the `endswith` changes might be a red herring, too, because I'm not seeing how the modified filter prevents the example given with `.py~`. 

Let's chat on this one, since this should really just be "script module extension blacklist", but we need other changes to make it actually behave that way.

**Comment by relrod** [comment]
> Controller plugin loaders should only allow *.py and not fall back to anything else. The discussion/comments about .pyc and .pyo being OK in certain contexts is generally wrong - the way the PluginLoader works today, we can basically never trust a binary Python module, because if it's bogus (eg wrong optimization level, wrong Python version), we're not going to go back looking for the source version, so things are just busted.

Sure. but we already blacklisted *.pyc and *.pyo in both of these places that this patch is touching. I'm not changing that or proposing changing that, just adding the rest of `BLACKLIST_EXTS` and a few extra to the default blacklist here.

> I think the `endswith` changes might be a red herring, too, because I'm not seeing how the modified filter prevents the example given with `.py~`.

It does. If have a collection with `a.py` and `a.py~` in it and I run a playbook that uses the `a` module, on `devel`, it will run the code in `a.py~`, and with this PR applied, it will run the code in `a.py`. Note that `~` is in `BLACKLIST_EXTS`.

**Review by bcoca — APPROVED** [review]

File: lib/ansible/config/base.yml [hunk 1]
```diff
@@ -1648,6 +1648,14 @@ INJECT_FACTS_AS_VARS:
   - {key: inject_facts_as_vars, section: defaults}
   type: boolean
   version_added: "2.5"
+MODULE_IGNORE_EXTS:
+  name: Module ignore extensions
+  default: "{{(BLACKLIST_EXTS + ('.yaml', '.yml', '.ini'))}}"
+  description: List of extensions to ignore when looking for modules to load
+  env: [{name: ANSIBLE_MODULE_IGNORE_EXTS}]
+  ini:
+  - {key: module_ignore_extensions, section: defaults}
```
Comment (nitzmahone) [inline lib/ansible/config/base.yml]: I'd go for consistency here with the envvar and actual config item name.

File: lib/ansible/config/base.yml [hunk 2]
```diff
@@ -1648,6 +1648,14 @@ INJECT_FACTS_AS_VARS:
   - {key: inject_facts_as_vars, section: defaults}
   type: boolean
   version_added: "2.5"
+MODULE_IGNORE_EXTS:
+  name: Module ignore extensions
+  default: "{{(BLACKLIST_EXTS + ('.yaml', '.yml', '.ini'))}}"
+  description: List of extensions to ignore when looking for modules to load
```
Comment (nitzmahone) [inline lib/ansible/config/base.yml]: Maybe just clarify in the description that this is for blacklisting script/binary module fallback extensions. I'd say change the name of the item to be clear that it's for script/binary module extensions to ignore, but I can't think of a reasonably concise name... `SCRIPT_BINARY_MODULE_IGNORE_EXTS`, while accurate and self-describing, feels too verbose.

File: test/integration/targets/plugin_loader/ignore_exts/library/demo.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/python
```
Comment (nitzmahone) [inline test/integration/targets/plugin_loader/ignore_exts/library/demo.py]: To really prove that this is working properly, you should probably also have something like an empty `demo.ini` in the same directory with this that would otherwise get picked up to prove that it kept looking.

File: test/sanity/ignore.txt [hunk 4]
```diff
@@ -315,6 +315,9 @@ test/integration/targets/pip/files/ansible_test_pip_chdir/__init__.py future-imp
 test/integration/targets/pip/files/ansible_test_pip_chdir/__init__.py metaclass-boilerplate
 test/integration/targets/pip/files/setup.py future-import-boilerplate
 test/integration/targets/pip/files/setup.py metaclass-boilerplate
+test/integration/targets/plugin_loader/ignore_exts/library/demo.py future-import-boilerplate
```
Comment (nitzmahone) [inline test/sanity/ignore.txt]: I think we're usually adding the boilerplate on new tests rather than adding new ignores

File: test/sanity/ignore.txt [hunk 5]
```diff
@@ -315,6 +315,9 @@ test/integration/targets/pip/files/ansible_test_pip_chdir/__init__.py future-imp
...
+test/integration/targets/plugin_loader/ignore_exts/library/demo.py metaclass-boilerplate
+test/integration/targets/plugin_loader/ignore_exts/library/demoignore.ini shebang
```
Comment (nitzmahone) [inline test/sanity/ignore.txt]: You could just put a bogus shebang in there to keep the test happy

File: test/integration/targets/plugin_loader/ignore_exts/ignore_exts.yml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+- hosts: testhost
```
Comment (nitzmahone) [inline test/integration/targets/plugin_loader/ignore_exts/ignore_exts.yml]: Should probably add some version of this test to the `collections` integration test target as well, since this is only testing the legacy side of it.

**Review by nitzmahone — CHANGES_REQUESTED** [review]

File: test/sanity/ignore.txt (hunk 4 above)
Comment (relrod) [reply test/sanity/ignore.txt]: Going to just do this differently and make it part of `module_precedence` tests instead.

**Review by nitzmahone — CHANGES_REQUESTED** [review]
We also usually try to keep the clones of existing real code for tests to a minimum, so a bunch of copies of the modules that shouldn't run is probably unncessary, and using "real" modules for these tests is also probably unnecessary. Take a look at the stub modules in the collections integration tests, eg https://github.com/ansible/ansible/blob/devel/test/integration/targets/collections/collection_root_user/ansible_collections/testns/testcoll/plugins/modules/ping.py- something like that or even just empty or bogus-shebang'd stuff for the ones that shouldn't run is probably sufficient (the difference being that a test failure might be a little harder to debug if the bogus modules aren't actually executable, so your call if you want to use actually executable stubs for the bogus ones or not). 

Otherwise changes LGTM- thanks!

**Comment by relrod** [comment]
I copied ping around because the existing code in that test did and I copied from them. But yeah we can use anything there.
